A general model for frictional contacts in colloidal systems

khofmann6 opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

I added a new class of pair force (similar to aniso). Here the relative "surface" velocity of two interacting particles is calculated (only for spherical particles) and is used to calculated a frictional force which may depend on this relative velocity. I added three frictional contact models (constant. linear, and Coulomb-Newton). 

For more details see the preprint: https://arxiv.org/abs/2507.16388

## Motivation and context

Frictional contacts couple translational and rotational degrees of freedom, which may affect the collective behavior of, e.g., colloids under shear and chiral active matter. Deterministic frictional contact models on the colloidal scale are thermodynamically inconsistent. This proposed model, which is roughly equivalent to DPD for anisotropic potentials, solves this inconsistency with a additional pairwise stochastic term. 

## How has this been tested?

The code was used to create the data presented in: https://arxiv.org/abs/2507.16388

pytest are still missing
Detailed documentation is missing

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [x] I have summarized these changes in `CHANGELOG.rst` following the established format.

Commit:
Added FrictionLJ pair interactions for frictional contacts

Commit:
Added name to credits

Commit:
Changed friction.py error

Commit:
Removed unnecessary imports in friction.py

Commit:
Changed order of classes to alphabetical order in friction.py

Commit:
Changed order of classes to alphabetical order in friction.py

joaander commented:
FYI, you can execute `pre-commit run --all-files` to automatically fix most of those errors. See [this page](https://hoomd-blue.readthedocs.io/en/v5.3.1/style.html) for more information.

Commit:
Changed order of classes to alphabetical order in friction.py

Commit:
pre-commit run applied

joaander commented:
Sorry, I've been very busy. I will review this pull request soon.

joaander commented on hoomd/md/EvaluatorPairFrictionLJBase.h:
```diff
@@ -0,0 +1,440 @@
+
+// $Id$
+// $URL$
+
```
```suggestion
```
Not necessary.

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+   is defined by \a param_type in the potential friction_evaluator class passed in. See the
+   appropriate documentation for the friction_evaluator for the definition of each element of the
+   parameters.
+*/
```
You don't need to repeat the comments of `PotentialPair` here. Reword the comment to state that `FrictionPair` is modeled after `PotentialPair` and state what is new and different.

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+        for (unsigned int i = 0; i < type_shape_mapping.size(); i++)
+            type_shapes.append(type_shape_mapping[i]);
+        return type_shapes;
+        }
```
I see that `EvaluatorPairLJBase` raises an error in `getShapeSpec()`.
```
throw std::runtime_error("Shape definition not supported for this pair potential.");
```

Do you plan for future frictional potentials that do implement it? If not, then you can remove `getTypeShapes` and `getTypeShapesPy`.

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                                                     unsigned int typ2,
+                                                     std::string action)
+    {
+    // TODO change logic to just throw an exception
```
```suggestion
```
It seems this TODO is already done.

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+
+/*! \param typ The type index.
+    \param param Shape parameter to set
+          set.
```
```suggestion
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+
+/*! \param typ The type index.
+    \param param Shape parameter to set
+          set.
```
```suggestion
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+
+/*! \param typ The type index.
+    \param param Shape parameter to set
+          set.
```
```suggestion
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                unsigned int j = h_nlist.data[myHead + k];
+                assert(j < m_pdata->getN() + m_pdata->getNGhosts());
+
+                // calculate dr_ji (MEM TRANSFER: 3 scalars / FLOPS: 3)
```
```suggestion
                // calculate dr_ji
```
I don't bother counting memory transfers and flops any more. Any performance is fine as long as the code performs well enough to achieve research goals.

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                // get the diameter of jth particle
+                Scalar dia_j = h_diameter.data[j];
+
+                // access the type of the neighbor particle (MEM TRANSFER: 1 scalar)
```
```suggestion
                // access the type of the neighbor particle
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+            const unsigned int size = (unsigned int)h_n_neigh.data[i];
+            for (unsigned int k = 0; k < size; k++)
+                {
+                // access the index of this neighbor (MEM TRANSFER: 1 scalar)
```
```suggestion
                // access the index of this neighbor
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                        }
+
+                    // add the force to particle j if we are using the third law (MEM TRANSFER: 10
+                    // scalars / FLOPS: 8)
```
```suggestion
                    // add the force to particle j if we are using the third law
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+        // for each particle
+        for (int i = 0; i < (int)m_pdata->getN(); i++)
+            {
+            // access the particle's position and type (MEM TRANSFER: 4 scalars)
```
```suggestion
            // access the particle's position and type
```

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                        torque_i.y = 0.0;
+                        torque_j.x = 0.0;
+                        torque_j.y = 0.0;
+                        }
```
This is now handled by the integrator: #2021

joaander commented on hoomd/md/FrictionPair.h:
```diff
@@ -0,0 +1,757 @@
+                    Scalar3 force2 = Scalar(0.5) * force;
+
+                    // add the force, potential energy and virial to the particle i
+                    // (FLOPS: 8)
```
```suggestion
```

joaander commented on hoomd/md/FrictionPairGPU.cuh:
```diff
@@ -0,0 +1,713 @@
+
+#include "hip/hip_runtime.h"
+// Copyright (c) 2009-2021 The Regents of the University of Michigan
+// This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
```
```suggestion
```

joaander commented on hoomd/md/pair/friction.py:
```diff
@@ -0,0 +1,133 @@
+
+r"""Frictional pair force classes apply a force, and torque on every
+particle in the simulation state.
+
```
```suggestion
Link to your paper here. Feel free to provide an more detailed overview of the method or not as you prefer.
```

joaander commented on hoomd/md/pair/friction.py:
```diff
@@ -0,0 +1,133 @@
+
+    `FrictionLJConstant` computes the frictional interaction
+    between pairs of particles.
+    """
```
You should add documentation for the type parameter dictionaries so that users know how to set the parameters.

Not required, but users strongly prefer to see example code blocks that demonstrate how to instantiate a class and set parameters. Place these in `.. code-block:: python` blocks and they will be executed automatically during testing. See #2110 for a recent example adding these blocks.

joaander commented on hoomd/md/EvaluatorPairFrictionLJBase.h:
```diff
@@ -0,0 +1,440 @@
+
+    struct shape_type
+        {
+        vec3<Scalar> mu;
```
Is this needed? It doesn't appear to be used anywhere. If you have no plans to support shapes in the future, you can remove the whole `shape_type` type, `setShape` and all related methods.

joaander reviewed (changes requested):
Thanks! I have some change requests and suggestions to improve this code.

You also need to fix the errors in the documentation build. [These instructions](https://hoomd-blue.readthedocs.io/en/v5.3.1/building.html#build-the-documentation) explain how to build the documentation locally. You can merge `trunk-minor` into your branch to fix some of the warnings, but there are several more warnings coming from your code.

Commit:
Merge remote-tracking branch 'upstream/trunk-minor' into trunk-patch

joaander commented:
You will also need to merge `trunk-minor` into this branch to use the updated CI infrastructure.

Commit:
Addressed the change requests and suggestions

khofmann6 committed:
Merge branch 'trunk-patch' into trunk-patch

Commit:
Applied style changes to friction.py

Commit:
Changed the example code for the user

Commit:
Fixed errors in FrictionPairGPU.cuh, FrictionPairGPU.h and FrictionPairFrictionLJBaseGPUKernel.cu that arised after removing everything related to shapes

Commit:
Fixed errors

Commit:
.

Commit:
.

Commit:
.

Commit:
Updated documentation

Commit:
Updated documentation

khofmann6 commented:
Thank you for the review. I addressed all requested changes and suggestions. Two quick questions:

1) I'm unsure if I removed the shape_type and set_shape related methods correctly. Could you review my removal of shape_type/set_shape, especially in the GPU code? 

For example in "FrictionPairGPU.cuh" i changed :

 // initialize extra shared mem
    char* s_extra = (char*)(s_shape_params + ntypes);

to 

 char* s_extra = (char*)( ntypes);

Im not sure if that is correct. My CUDA knowledge is limited, and I want to be sure it's right. 

2) I'm still receiving errors for the documentation, and I can't figure them out.

Thank you for your support.

joaander committed:
Fix doc build errors with inspect.cleandoc.

joaander committed:
Apply suggestions from code review.

joaander committed:
Wrap aligned equations in align blocks.

joaander commented:
Thanks for the updates! I've pushed a few comments that apply my other suggestions directly.

>     1. I'm unsure if I removed the shape_type and set_shape related methods correctly. Could you review my removal of shape_type/set_shape, especially in the GPU code?
> 
> 
> For example in "FrictionPairGPU.cuh" i changed :
> 
> // initialize extra shared mem char* s_extra = (char*)(s_shape_params + ntypes);
> 
> to
> 
> char* s_extra = (char*)( ntypes);
> 
> Im not sure if that is correct. My CUDA knowledge is limited, and I want to be sure it's right.

This should be:
```
    char* s_extra = (char*)(s_params + num_typ_parameters);
```
so that the extra shared memory doesn't overlap with the other type parameters.
 
>     2. I'm still receiving errors for the documentation, and I can't figure them out.

I pushed a commit that calls `inspect.cleandoc` to fix this.

joaander committed:
Set non-overlapping s_extra block.

Commit:
Changed name of constant friction model to Coulomb friction model

joaander committed:
Remove unused shift_mode.

joaander committed:
Fix compile error.

joaander committed:
Correctly set one accepted mode.

joaander committed:
Mention pair.friction.

joaander reviewed (changes requested):
Thanks! This is almost ready to merge.

I made changes to fix the GPU kernels. Please run your external tests and make sure that the code still behaves correctly.

You should also add a pytest unit test. You can't validate the exact values of the stochastic forces and torques, but you should still provide a tests that use the various pair potentials and ensure that that they can run a few timesteps. Choose the parameters of the initial condition so that the particles interact (via the `d` parameters of the `two_particle_snapshot_factory`) and you can at least verify a non-zero force and torque.

Commit:
Changed a typo in friction.py and added the file test_friction.py which tests the setting of the parameter, the deterministic friction force and torque, and if the complete friction interaction runs for a few timesteps

khofmann6 committed:
Merge branch 'trunk-patch' into trunk-patch

Commit:
.

khofmann6 commented:
Thank you!

 I added a test_friction-py file which test one two particle configuration for each friction model. Here I check the exact values for force and torque in a deterministic case and if the simulation is able to run for a few time steps. But I need to run my external tests to make sure that the code still behaves correctly.

Commit:
Added blank line in CHANGELOG

Commit:
N

khofmann6 commented:
As a test, I used the new GPU code to recreate some old figures. The new code behaves correctly and produces consistent results.

Thank you for your help!

joaander reviewed (approved):
Thanks!
